File: apps/web/src/types/seller.ts
import type { Tool } from "@/types/tool";

export interface SellerRevenuePoint {
  date: string;
  amount: string;
}

export interface SellerRequestsPoint {
  date: string;
  count: number;
}

export interface SellerLatencyPoint {
  date: string;
  avg_response_time_ms: number;
}

export interface SellerActivityItem {
  id: string;
  tool_id: string;
  tool_name: string;
  request_timestamp: string;
  status_code: number;
  response_time_ms: number;
  cost: string;
  error_message: string | null;
}

export interface SellerToolSummary {
  tool_id: string;
  tool_name: string;
  slug: string;
  status: "draft" | "processing" | "live" | "paused" | "rejected";
  latest_job_status: ToolProcessingJobStatus | null;
  latest_job_error: string | null;
  requests_this_month: number;
  revenue_this_month: string;
  unique_users_this_month: number;
  avg_response_time_ms: number | null;
  p50_response_time_ms: number | null;
  p95_response_time_ms: number | null;
  p99_response_time_ms: number | null;
  uptime_percentage: string | null;
}

export interface SellerTopTool {
  tool_id: string;
  tool_name: string;
  revenue_this_month: string;
}

export interface SellerDashboardResponse {
  total_tools: number;
  total_revenue_all_time: string;
  total_revenue_this_month: string;
  previous_month_revenue: string;
  total_requests_this_month: number;
  active_tools: number;
  avg_response_time_ms: number | null;
  top_tool: SellerTopTool | null;
  revenue_chart_data: SellerRevenuePoint[];
  request_chart_data: SellerRequestsPoint[];
  latency_chart_data: SellerLatencyPoint[];
  recent_activity: SellerActivityItem[];
  tools: SellerToolSummary[];
}

export interface SellerErrorSummary {
  error_message: string;
  count: number;
}

export interface SellerErrorLogItem {
  timestamp: string;
  error_message: string | null;
  status_code: number;
  input_size_bytes: number;
  output_size_bytes: number;
  response_time_ms: number;
}

export interface SellerAnalyticsResponse {
  period: "7d" | "30d" | "90d" | "all";
  requests_over_time: SellerRequestsPoint[];
  revenue_over_time: SellerRevenuePoint[];
  unique_users: number;
  avg_response_time_ms: number | null;
  error_rate: number;
  top_errors: SellerErrorSummary[];
  geographic_distribution: Array<Record<string, string | number>>;
  p50_response_time_ms: number | null;
  p95_response_time_ms: number | null;
  p99_response_time_ms: number | null;
  recent_errors: SellerErrorLogItem[];
}

export type ToolProcessingJobStatus =
  | "queued"
  | "running"
  | "retrying"
  | "succeeded"
  | "failed";

export interface ToolProcessingJob {
  id: string;
  tool_id: string;
  status: ToolProcessingJobStatus;
  attempts: number;
  max_attempts: number;
  trigger: string;
  last_error: string | null;
  arq_job_id: string;
  enqueued_at: string | null;
  started_at: string | null;
  finished_at: string | null;
  created_at: string;
  updated_at: string;
}

export interface SellerSubmissionStatusResponse {
  tool: Tool;
  job: ToolProcessingJob | null;
}

File: apps/api/app/schemas/seller.py
import uuid
from datetime import date, datetime
from decimal import Decimal

from pydantic import BaseModel, ConfigDict

from app.models.tool import ToolStatus
from app.models.tool_processing_job import ToolProcessingJobStatus
from app.schemas.tool import ToolResponse


class SellerRevenuePoint(BaseModel):
    date: date
    amount: Decimal


class SellerRequestsPoint(BaseModel):
    date: date
    count: int


class SellerToolSummary(BaseModel):
    tool_id: uuid.UUID
    tool_name: str
    slug: str
    status: ToolStatus
    latest_job_status: ToolProcessingJobStatus | None = None
    latest_job_error: str | None = None
    requests_this_month: int
    revenue_this_month: Decimal
    avg_response_time_ms: float | None = None


class SellerTopTool(BaseModel):
    tool_id: uuid.UUID
    tool_name: str
    revenue_this_month: Decimal


class SellerDashboardResponse(BaseModel):
    total_tools: int
    total_revenue_all_time: Decimal
    total_revenue_this_month: Decimal
    previous_month_revenue: Decimal
    total_requests_this_month: int
    active_tools: int
    avg_response_time_ms: float | None = None
    top_tool: SellerTopTool | None = None
    revenue_chart_data: list[SellerRevenuePoint]
    tools: list[SellerToolSummary]


class ToolProcessingJobResponse(BaseModel):
    model_config = ConfigDict(from_attributes=True)

    id: uuid.UUID
    tool_id: uuid.UUID
    status: ToolProcessingJobStatus
    attempts: int
    max_attempts: int
    trigger: str
    last_error: str | None = None
    arq_job_id: str
    enqueued_at: datetime | None = None
    started_at: datetime | None = None
    finished_at: datetime | None = None
    created_at: datetime
    updated_at: datetime


class SellerSubmissionStatusResponse(BaseModel):
    tool: ToolResponse
    job: ToolProcessingJobResponse | None = None


class SellerErrorSummary(BaseModel):
    error_message: str
    count: int


class SellerErrorLogItem(BaseModel):
    timestamp: datetime
    error_message: str | None = None
    status_code: int
    input_size_bytes: int
    output_size_bytes: int
    response_time_ms: int


class SellerAnalyticsResponse(BaseModel):
    period: str
    requests_over_time: list[SellerRequestsPoint]
    revenue_over_time: list[SellerRevenuePoint]
    unique_users: int
    avg_response_time_ms: float | None = None
    error_rate: float
    top_errors: list[SellerErrorSummary]
    geographic_distribution: list[dict[str, str | int]]
    p50_response_time_ms: float | None = None
    p95_response_time_ms: float | None = None
    p99_response_time_ms: float | None = None
    recent_errors: list[SellerErrorLogItem]
